rfcs/build-system/02_standard_package_output_folders.md

```diff
@@ -0,0 +1,86 @@
+# RFC: Standard package output folders
+
+---
+
+_List contributors to the proposal: @dzearing_
+
+## Summary
+
+There are multiple javascript formats we may include in our npm packages: `commonjs`, `esm`, and `amd`, in addition to static assets like pre-made bundles and images.
+
+We should adopt a predictable folder structure within the published npm package for these formats. That way consumers of our packages can know exactly where things are and which folders have what things in as few steps as possible.
+
+This is one of those small details that other partners will emulate and adopt. If we can be consistent, many packages exported by our partners will likely follow guidance because it is really minutia, but ends up adding to more predictability at scale.
+
+## Problem statement
+
+Today, `lib` contains esm JavaScript output, except for node-only packages which will drop CommonJS modules in them. This obfuscates what `lib` actually contains. We should adopt standards which add clarity to what output format is used.
+
+## Detailed Design or Proposal
+
+We've historically used the following standard output folders in a published JavaScript package (only applicable output folders would be present):
+
+- `/lib` - esm (as we've had for a long time)
+- `/lib-commonjs` - commonjs (only needed while Node <= 13.2.0 is supported)
+- `/lib-amd` - amd (hopefully we can drop someday)
+- `/dist` - bundles and static content
```

Comment:
_My personal preference is to have `lib-esm` or `lib-es` as it obviously says what is in a directory._

---

Another thing that we discussed some time ago to have a single `dist` folder. For example:

- `/dist/esm` - esm (as we've had for a long time)
- `/dist/commonjs` - commonjs (only needed while Node <= 13.2.0 is supported)
- `/dist/amd` - amd (hopefully we can drop someday)
- `/dist/assets` - bundles and static content

It will have the same props&cons as an alternative proposal, but it will be easier to ignore a single entry in `.gitignore`/IDE 🐱

Reply:
If we want to change things, I agree this would be even cleaner

Reply:
Also I could behind the `lib-esm` approach. It's more explicit, and doesn't favor a particular output. This means we can add more output flavors in the future (the react-native output formats would be the first which come to mind) without the akwardness of picking one as the "true" output flavor.

My primary concern changing now is to avoid destabilizing partners, as Elizabeth mentioned below. My recommendation would be to stick with the current output folders in the writeup, fix the node packages only to stay inline with other packages (since this was a recent change.)

Then in the converged packages, we could start adopting the dist/* layout now.  Later in v9+, change everything to be inline with the dist approach.

Is this reasonable/ideal? Concerns?

Reply:
> My primary concern changing now is to avoid destabilizing partners, as Elizabeth mentioned below. My recommendation would be to stick with the current output folders in the writeup, fix the node packages only to stay inline with other packages (since this was a recent change.)
> 
> Then in the converged packages, we could start adopting the dist/* layout now. Later in v9+, change everything to be inline with the dist approach.
> 
> Is this reasonable/ideal? Concerns?

No concerns, make sense for me 👍 

Should we update a proposed solution to include v8 (as it stays currently) and converged proposal (for example, `/dist` layout, or another proposal) in this RFC? Or have an another one?

Reply:
I think it's simple enough that we could cover both v8 and converged in this RFC. 

Though the part we didn't totally decide on for v8 is whether `lib` *must* be ESM for *all* packages, even ones where in reality we only need commonjs.